sql/src/main/java/io/crate/metadata/RelationName.java

```diff
@@ -45,8 +48,42 @@
     private final String schema;
     private final String name;
 
-    public static RelationName of(Table tableNode, String defaultSchema) {
-        return of(tableNode.getName(), defaultSchema);
+    public static RelationName resolveRelation(Table tableNode, SearchPath searchPath, Schemas schemas) {
+        return resolveRelation(tableNode.getName(), searchPath, schemas);
+    }
+
+    public static RelationName resolveRelation(QualifiedName name, SearchPath searchPath, Schemas schemas) {
```

Comment:
I'm not sure if it's nice to have the `RelationName` become aware of lookup logic and of table/view semantics. That we're looking up objects which are discarded / not used is also a bit fishy.

Maybe we should rework the public methods on `Schemas` to better reflect the new relation resolving logic?

Reply:
I'm not sure about this. I pondered about this as well.

What do you mean by "we're looking up objects which are discarded / not used is also a bit fishy"? Looking up in `Schemas`?

Views and tables are relations so why shouldn't the `RelationName` know about it? I think the logic of "let's see what relation are we really looking for"  is currently encapsulated in the entity/factory that identifies these relations. 

If we move this somewhere else we'll probably have to change the meaning of a `RelationName` (schema should be optional for eg) or whoever's creating/resolving the `RelationNames` should work with optional schema and a string identifying the relation name or something similar.

Reply:
> What do you mean by "we're looking up objects which are discarded / not used is also a bit fishy"? Looking up in Schemas?

I meant that it is retrieving TableInfo instances and ViewMetaData instances which are never used.

Most call sites look something like this:

```
DocTableInfo docTableInfo = schemas.getTableInfo(
            RelationName.resolveRelation(table, sessionContext.searchPath(), schemas))
```

If you'd "inline" the logic, it would look rather weird. (Fetching
`DocTableInfo` instances, drop them to build the `relationName` to again fetch
a `DocTableInfo`. In the current implementation the `RelationName` that was built to fetch a `TableInfo` instances is dropped, just to mark the schema and then build the very same `RelationName` again.

Also the signature itself: `schemas.getTableInfo(..., schemas)` is a bit strange.


I think this all points to `Schemas` being the better place to encapsulate the
logic.

E.g. we could have quick methods by-passing the `RelationName` altogether. Like
`getTableInfo(QualifiedName, SearchPath)`. And maybe some
`findMatchingRelation(QualifiedName, SearchPath) -> RelationName` ? Depending on the use-sites.

The `RelationName` constructors could probably be kept as they are.